drivers/dma/dma_esp32_gdma.c

```diff
@@ -110,23 +125,58 @@ static void IRAM_ATTR dma_esp32_isr_handle(const struct device *dev, uint8_t rx_
 	}
 }
 
+#if defined(CONFIG_SOC_ESP32C3)
+static int dma_esp32_enable_interrupt(const struct device *dev,
+				      struct dma_esp32_channel *dma_channel)
+{
+	struct dma_esp32_config *config = (struct dma_esp32_config *)dev->config;
+
+	return esp_intr_enable(config->irq_src[dma_channel->channel_id]);
+}
+
+static int dma_esp32_disable_interrupt(const struct device *dev,
+				       struct dma_esp32_channel *dma_channel)
+{
+	struct dma_esp32_config *config = (struct dma_esp32_config *)dev->config;
+
+	return esp_intr_disable(config->irq_src[dma_channel->channel_id]);
+}
+#else
+static int dma_esp32_enable_interrupt(const struct device *dev,
+				      struct dma_esp32_channel *dma_channel)
+{
+	struct dma_esp32_config *config = (struct dma_esp32_config *)dev->config;
+
+	return esp_intr_enable(dma_channel->intr_handle);
+}
+
+static int dma_esp32_disable_interrupt(const struct device *dev,
+				       struct dma_esp32_channel *dma_channel)
+{
+	struct dma_esp32_config *config = (struct dma_esp32_config *)dev->config;
+
+	return esp_intr_disable(dma_channel->intr_handle);
+}
+
+#endif
 static int dma_esp32_config_rx_descriptor(struct dma_esp32_channel *dma_channel,
 					   struct dma_block_config *block)
 {
 	memset(&dma_channel->desc, 0, sizeof(dma_channel->desc));
 	dma_channel->desc.buffer = (void *)block->dest_address;
 	dma_channel->buf_original = (uint8_t *)block->dest_address;
+	k_free(dma_channel->buf_temp);
+	dma_channel->buf_temp = NULL;
 	if (!esp_ptr_dma_capable((uint32_t *)block->dest_address)) {
```

Comment:
Why are we trying to allocate on behalf of callers here, if an invalid buffer for DMA is being provided this seems like a bug that should be asserted on.

I'm a little confused by the code here to malloc/free a temporary buffer when that's not the case. Is there a particular scenario esp32 expects this to occur?

Reply:
We are freeing the channel's temporary buffer every time a new descriptor needs to be created. In case the original buffer is in a valid DMA region, the `k_free` can handle a NULL pointer without issue and no allocation is needed. In case it is in an invalid region, we create a temporary one to handle the transactions, freeing it and reallocating according the size required by the transactions. I'm trying to deal with both scenarios here, since the original buffer location depends on the application or driver using it.  

Did I answer your question? I'm not shure haha

Reply:
I guess what I'm suggesting is this is a bug then in the driver or application that should be caught. There's an RFC for managing buffers https://github.com/zephyrproject-rtos/zephyr/pull/57602

At the moment this would be, to my best knowledge, the only DMA driver that attempts to fix the buffer you asked a copy to by allocating from the heap. I would suggest pushing that issue up the stack as this is a surprising place to have that happen.

Reply:
Oh nice! I will follow up this RFC and contribute in whatever way I can. But since it is an ongoing discussion, should we block the merge of this commit? This strategy of dealing with buffer is around for quite a while, this commit only  add support for esp32s3 and  small fixes.

Reply:
I'd point out that generally Zephyr is trying to follow misra, and in misra-c "dynamic allocation is not allowed" See our guidelines https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#main-rules

So I think this is something that needs to be revisited but today should be considered a programming error trying to ask DMA to work with a buffer that doesn't conform to DMA requirements.